lib/puppet/provider/iis_feature/default.rb

```diff
@@ -0,0 +1,157 @@
+require File.join(File.dirname(__FILE__), '../../../puppet/provider/iis_powershell')
+
+Puppet::Type.type(:iis_feature).provide(:default, parent: Puppet::Provider::IIS_PowerShell) do
+  desc "IIS feature provider"
+
+  confine    :operatingsystem  => [ :windows ]
+  defaultfor :operatingsystem  => :windows
+
+  commands :powershell => PuppetX::IIS::PowerShellCommon.powershell_path
+
+  mk_resource_methods
+
+  def exists?
+    @property_hash[:ensure] == :present
+  end
+
+  def create
+    if iis_installable_features.includes?(@resource[:name])
+      raise Puppet::Error, "iis_feature can only be used to install IIS features. '#{resource[:name]}' is not an IIS feature"
+    end
+
+    if @resource[:include_management_tools] == true && self.class.is_windows2008 == true
+      raise Puppet::Error, 'include_management_tools can only be used with Windows 2012 and above'
+    end
+
+    cmd = []
+    cmd << "Import-Module ServerManager; Add-WindowsFeature #{resource[:name]}" if self.class.is_windows2008 == true
+    cmd << "Install-WindowsFeature #{resource[:name]}" if self.class.is_windows2008 == false
+    cmd << '-IncludeAllSubFeature' if @resource[:include_all_subfeatures] == true
+    cmd << '-Restart' if @resource[:restart] == true
+    cmd << "-Source #{resource['source']}" if @resource[:source] == false
+    cmd << '-IncludeManagementTools' if @resource[:include_management_tools] == true && self.class.is_windows2008 == false
+
+    Puppet.debug "Powershell create command is '#{cmd}'"
+    result = self.class.run(cmd.join)
+    Puppet.debug "Powershell create response was '#{result}'"
+  end
+
+  def destroy
+    if iis_installable_features.includes?(@resource[:name])
+      raise Puppet::Error, "iis_feature can only be used to uninstall IIS features. '#{resource[:name]}' is not an IIS feature"
+    end
+
+    cmd = []
+    cmd << "Import-Module ServerManager; Remove-WindowsFeature #{resource[:name]}" if self.class.is_windows2008 == true
+    cmd << "Uninstall-WindowsFeature #{resource[:name]}" if self.class.is_windows2008 == false
+    cmd << '-Restart' if @resource[:restart] == true
+
+    Puppet.debug "Powershell destroy command is '#{cmd}'"
+    result = self.class.run(cmd.join)
+    Puppet.debug "Powershell destroy response was '#{result}'"
+  end
+
+  def self.instances
+    cmd = []
+    cmd << 'Import-Module ServerManager; ' if self.is_windows2008 == true
+    cmd << 'Get-WindowsFeature | ConvertTo-Json -Depth 4'
+    
+    result = self.run(cmd.join)
+    
+    return [] if result.nil?
+
+    json = self.parse_json_result(result[:stdout])
+    return [] if json.nil?
+
+    return json.collect do |feature|
```

Comment:
Ruby implicitly returns the last result in a method so this return isn't needed, but it's fine to have it. ruby-lint might warn about this.